### PR TITLE
feat: add vision and image generation tools

### PR DESCRIPTION
**Pull request by njhale**
Add an `Images` tool bundle with tools for image generation and analysis.

Note: Open to suggestions on better names for this bundle.

### CODE REVIEW

File: images/src/tools.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                process.env.PROMPT,
+                // Split the string into an array of image URLs, while being careful to handle URLs that contain commas.
+                process.env.IMAGES?.split(/(?<!\\),/).map(image => image.replace(/\\,/g, ',')),
+                parseInt(process.env.MAX_TOKENS ?? '2048'),
+            )
+            break
+        case 'generateImages':
+            generateImages(
+                process.env.PROMPT,
+                process.env.MODEL,
```
Comment (g-linville) [inline images/src/tools.ts:24]: It doesn't look this this is exposed as a `Param:` for this tool. Should we add it?

File: images/src/generate.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import OpenAI from 'openai';
+
+type ImageSize = '1024x1024' | '256x256' | '512x512' | '1792x1024' | '1024x1792';
```
Comment (g-linville) [inline images/src/generate.ts:3]: If these are the only size options, I think we should do more validation around this and also let the model know in the parameter description that it has to be one of these sizes.

File: images/src/generate.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+
+  const openai = new OpenAI();
+
+  try {
+    const response = await openai.images.generate({
+      model,
+      prompt,
+      size: size as ImageSize,
```
Comment (g-linville) [inline images/src/generate.ts:24]: (Specifically here is where I think we need to do validation, unless this throws an exception if `size` has no corresponding `ImageSize`. I don't know if that's what happens.)

File: images/src/tools.ts (hunk 1 above)
Comment (njhale) [reply images/src/tools.ts:24]: Not sure users should be able to change the model yet since some models don't support vision.

Comment (drpebcak) [reply images/src/tools.ts:24]: There's decent vision support out there, its really image-generation that's lacking.

Without a param for model, this is a regression for our support of vision with the ollama-provider.

File: images/src/analyze.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      image_url: {
+        detail: 'auto',
+        url: await resolveImageURL(image),
+      },
+    }))
+  ) as ChatCompletionContentPartImage[];
+
+  const openai = new OpenAI();
+  const response = await openai.chat.completions.create({
+    model: 'gpt-4o',
```
Comment (drpebcak) [inline images/src/analyze.ts:51]: Not sure we want this hardcoded. There are other models that can support vision if we are more flexible.

**Review by drpebcak — CHANGES_REQUESTED** [review]

File: images/src/tools.ts (hunk 1 above)
Comment (njhale) [reply images/src/tools.ts:24]: Got it. I'll add the model param back to the tool.

File: images/src/generate.ts (hunk 2 above)
Comment (njhale) [reply images/src/generate.ts:3]: done

File: images/src/generate.ts (hunk 3 above)
Comment (njhale) [reply images/src/generate.ts:35]: done

File: images/src/tools.ts (hunk 1 above)
Comment (njhale) [reply images/src/tools.ts:23]: done

File: images/src/analyze.ts (hunk 4 above)
Comment (njhale) [reply images/src/analyze.ts]: I plumbed this down from the new param.

**Review by thedadams — APPROVED** [review]
This looks good. However, I do remember a request to write tools in Python rather than Javascript. Not sure if that makes sense here.

**Comment by njhale** [comment]
> This looks good. However, I do remember a request to write tools in Python rather than Javascript. Not sure if that makes sense here.

@thedadams I had a conversation with Craig about this right before he went on vacation. The gist of it was to use whatever language makes the most sense for the tool we're working on. In this case, I felt like typescript was a good fit because other tools in this repo use it and vision was already a nodejs app.

I can convert this to python if you feel strongly about it.

**Comment by thedadams** [comment]
@njhale Doesn't matter to me.

**Review by drpebcak — APPROVED** [review]

**Pull request merged**